src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php

```diff
@@ -0,0 +1,191 @@
+     */
+    private $sessionManager;
+
+    /**
+     * @var AccessControlQueryEnhancer
+     */
+    private $accessControlQueryEnhancer;
+
+    /**
+     * @var Security|null
+     */
+    private $security;
+
+    /**
+     * @var array<string, int>
+     */
+    private $permissions;
+
+    /**
+     * @param array<string, int> $permissions
+     */
+    public function __construct(
+        SessionManagerInterface $sessionManager,
+        AccessControlQueryEnhancer $accessControlQueryEnhancer,
+        ?Security $security,
+        array $permissions
+    ) {
+        $this->sessionManager = $sessionManager;
+        $this->accessControlQueryEnhancer = $accessControlQueryEnhancer;
+        $this->security = $security;
+        $this->permissions = $permissions;
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            Events::REMOVE => ['handleRemove', 4096],
+            Events::CONFIGURE_OPTIONS => ['handleConfigureOptions'],
+        ];
+    }
+
+    /**
+     * @param array<array{id: string, resourceKey: string, depth: int}> $resources
+     *
+     * @return array<int, array<array{id: string, resourceKey: string}>>
+     */
+    private function groupResourcesByDepth(array $resources): array
+    {
+        $grouped = [];
+
+        foreach ($resources as $resource) {
+            $depth = $resource['depth'];
+            unset($resource['depth']);
+
+            if (!isset($grouped[$depth])) {
+                $grouped[$depth] = [];
+            }
+
+            $grouped[$depth][] = $resource;
+        }
+
+        \krsort($grouped);
+
+        return \array_values($grouped);
+    }
+
+    public function handleRemove(RemoveEvent $event): void
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof BasePageDocument) {
+            return;
+        }
+
+        $forceRemoveChildren = (bool) $event->getOption(static::FORCE_REMOVE_CHILDREN_OPTION, false);
+
+        if ($forceRemoveChildren) {
+            return;
+        }
+
+        $query = $this->createSql2Query(
+            \sprintf(
+                'SELECT [jcr:uuid] FROM [nt:unstructured] AS page WHERE [jcr:mixinTypes] = "sulu:page" AND ISDESCENDANTNODE(page, "%s")',
+                $document->getPath()
+            )
+        );
+        $result = $query->execute();
+
+        $descendantPages = [];
+
+        /** @var RowInterface<mixed> $row */
+        foreach ($result->getRows() as $row) {
+            $descendantPages[] = [
+                'id' => $row->getValue('page.jcr:uuid'),
+                'depth' => \substr_count((string) $row->getPath(), '/'),
+                'resourceKey' => PageDocument::RESOURCE_KEY,
+            ];
+        }
+
+        $descendantPageIds = \array_column($descendantPages, 'id');
+        $descendantAuthorizedPageIds = $this->accessControlQueryEnhancer->findIdsWithGrantedPermissions(
```

Comment:
this function in the enhancer is strange - you have mentioned it already - i would move it into a repository